File: part_10-model-selection-and-boosting/model_selection/grid_search.py
# K-Fold Cross Validation

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('../../data_files/Social_Network_Ads.csv')
X = dataset.iloc[:, [2, 3]].values
y = dataset.iloc[:, 4].values

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Fitting classifier to the Training set
from sklearn.svm import SVC
classifier = SVC(kernel='rbf', random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# Applying K-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)
accuracies.mean()
accuracies.std()

# Applying grid search to find the best model and best parameters
from sklearn.model_selection import GridSearchCV
parameters = [
    {
        'c': [1, 10, 100, 1000],
        'kernel': ['linear']
    },
    {
        'c': [1, 10, 100, 1000],
        'kernel': ['rbf'],
        'gamma': [i*0.1 for i in range(10)]
    }
]
grid_search = GridSearchCV(estimator=classifier, param_grid=parameters, scoring='accuracy', cv=10, n_jobs=-1)
grid_search = grid_search.fit(X_train, y_train)
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_

File: part_10-model-selection-and-boosting/model_selection/grid_search.R
# Importing the dataset
# ---------------------
dataset = read.csv('../../data_files/Social_Network_Ads.csv')
dataset = dataset[, 3:5]

# Splitting the dataset into the Training set and Test set
# --------------------------------------------------------
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split==TRUE)
test_set = subset(dataset, split==FALSE)

# Feature scaling
# ---------------
training_set[, 1:2] = scale(training_set[, 1:2])
test_set[, 1:2] = scale(test_set[, 1:2])

# Fitting Classifier to training set
#---------------------------------------------
library(e1071)
classifier = svm(formula=Purchased ~ .,
                 data=training_set,
                 type='C-classification',
                 kernel='radial')

# Predicting the Test set results
#--------------------------------
y_pred = predict(classifier, newdata=test_set[-3])

# Making the Confusion Matrix
cm = table(test_set[, 3], y_pred)

# Applying k-Fold Cross Validation
library(caret)
folds = createFolds(training_set$Purchased, k=10)
cv = lapply(folds, function(x) {
    training_fold = training_set[-x, ]
    test_fold = training_set[x, ]
    classifier = svm(formula=Purchased ~ .,
                     data=training_fold,
                     type='C-classification',
                     kernel='radial')
    y_pred = predict(classifier, newdata=test_fold[-3])
    cm = table(test_fold[, 3], y_pred)
    accuracy = (cm[1, 1] + cm[2, 2]) / (cm[1, 1] + cm[2, 2] + cm[1, 2] + cm[2, 1])
    return(accuracy)
})
accuracy = mean(as.numeric(cv))

# Applying Grid Search to find the best parameters
library(caret)
classifier = train(form=Purchased ~ ., data=training_set, method='svmRadial')
classifier$bestTune
